### PR TITLE
Fix various warnings

### PR DESCRIPTION
**Pull request by ZedThree**
Fixes the following warnings on gcc:
- overloaded-virtual
- old-style-cast
- non-virtual-dtor
- pedantic

There are >300 warnings from our use of `int` instead of `size_t` in lots of places. I'm not a huge fan of unsigned types, but seeing as the standard library does use them, we might be better off switching to use them. There's probably a few places where it's not as simple as just changing the type, and we'd have to make other changes, but the majority of places are likely fine.

### CODE REVIEW

File: include/bout/field_visitor.hxx [hunk 1]
```diff
@@ -22,7 +22,7 @@ public:
   virtual void accept(Field3D &f) = 0;
   virtual void accept(Vector2D &f) = 0;
   virtual void accept(Vector3D &f) = 0;
-  
+  virtual ~FieldVisitor() = default;
 };
```
Comment (bendudson) [inline include/bout/field_visitor.hxx]: semicolon?

File: include/field.hxx [hunk 2]
```diff
@@ -505,22 +505,22 @@ inline T pow(BoutReal lhs, const T& rhs, REGION rgn) {
 /// This loops over the entire domain, including guard/boundary cells by
 /// default (can be changed using the \p rgn argument).
 /// If CHECK >= 3 then the result will be checked for non-finite numbers
-FIELD_FUNC(sqrt, ::sqrt);
+FIELD_FUNC(sqrt, ::sqrt)
```
Comment (bendudson) [inline include/field.hxx]: I think I may have put in a lot of these semicolons after macro calls. I think it was due to the editor indentation; without the semicolon my editor at the time would think the next line was a continuation, and helpfully indent it. Is this an issue for clang formatting?

File: include/bout/field_visitor.hxx (hunk 1 above)
Comment (ZedThree) [reply include/bout/field_visitor.hxx]: Required here :)

Comment (bendudson) [reply include/bout/field_visitor.hxx]: I am so confused...

File: include/field.hxx (hunk 2 above)
Comment (ZedThree) [reply include/field.hxx]: Yeah, I seem to recall having problems here in the past, but both emacs and clang-format Do The Right Thing now, so I'm happy removing them.

File: include/bout/field_visitor.hxx (hunk 1 above)
Comment (ZedThree) [reply include/bout/field_visitor.hxx]: It's the end of a `class` declaration

Comment (bendudson) [reply include/bout/field_visitor.hxx]: but the end of a template class doesn't need one?

**Pull request merged**